File: orcastork_lite/state.py
"""``SessionState`` — the session's in-memory DataPoint set, revisions and watermarks.

Keyed-merge on ``(class, value)``: an equal DataPoint **merges** into the existing entry
(keeping ``first_retrieved``, advancing ``last_retrieved``) rather than duplicating. Every
batch that changes something advances one monotonic revision, and each entry remembers the
revision it was added at and last freshened at; an operator's **watermark** is the revision its
last run observed, so the delta since then splits cleanly into added vs updated.
"""

from __future__ import annotations

from collections.abc import Iterable
from typing import Any

from pydantic import BaseModel, ConfigDict

from .datapoints import DataPoint, DataPointView, identity_key
from .ids import CapabilityId, OperatorId
from .operators import InvocationDelta


class _Entry(BaseModel):
    data_point: DataPoint  # bare: see InvocationDelta for why the annotation is not parametrized
    added_rev: int
    updated_rev: int


class MergeOutcome(BaseModel):
    """What one ``merge`` did: the resulting revision and which DataPoints were new vs freshened."""

    model_config = ConfigDict(frozen=True)

    revision: int
    added: tuple[DataPoint, ...]
    updated: tuple[DataPoint, ...]

    @property
    def changed(self) -> bool:
        return bool(self.added or self.updated)


class SessionState:
    def __init__(self) -> None:
        self._entries: dict[tuple[type[DataPoint[Any]], Any], _Entry] = {}
        self._revision = 0
        self._watermarks: dict[OperatorId, int] = {}

    @property
    def revision(self) -> int:
        return self._revision

    def merge(self, data_points: Iterable[DataPoint[Any]]) -> MergeOutcome:
        """Keyed-merge a batch; the revision advances iff something was added or freshened."""
        next_revision = self._revision + 1
        added: list[DataPoint[Any]] = []
        updated: list[DataPoint[Any]] = []
        for data_point in data_points:
            key = identity_key(data_point)
            existing = self._entries.get(key)
            if existing is None:
                self._entries[key] = _Entry(data_point=data_point, added_rev=next_revision, updated_rev=next_revision)
                added.append(data_point)
            elif data_point.last_retrieved > existing.data_point.last_retrieved:
                existing.data_point = existing.data_point.reobserved(data_point.last_retrieved)
                existing.updated_rev = next_revision
                updated.append(existing.data_point)
        if added or updated:
            self._revision = next_revision
        return MergeOutcome(revision=self._revision, added=tuple(added), updated=tuple(updated))

    def view(self) -> DataPointView:
        return DataPointView(entry.data_point for entry in self._entries.values())

    def has_run(self, operator_id: OperatorId) -> bool:
        return operator_id in self._watermarks

    def advance_watermark(self, operator_id: OperatorId, revision: int) -> None:
        self._watermarks[operator_id] = revision

    def delta_for(
        self,
        operator_id: OperatorId,
        *,
        previous_caps: frozenset[CapabilityId],
        available_caps: frozenset[CapabilityId],
    ) -> InvocationDelta:
        """What changed since ``operator_id`` last ran; a first run presents the whole set as ``added``."""
        newly_available = available_caps - previous_caps
        watermark = self._watermarks.get(operator_id)
        entries = self._entries.values()
        if watermark is None:
            return InvocationDelta(
                added=frozenset(entry.data_point for entry in entries),
                updated=frozenset(),
                newly_available_caps=newly_available,
                is_first_invocation=True,
            )
        return InvocationDelta(
            added=frozenset(entry.data_point for entry in entries if entry.added_rev > watermark),
            updated=frozenset(
                entry.data_point for entry in entries if entry.added_rev <= watermark < entry.updated_rev
            ),
            newly_available_caps=newly_available,
            is_first_invocation=False,
        )

File: typescript/src/orcastork_lite/state.ts
/**
 * `SessionState` — the session's in-memory DataPoint set, revisions and watermarks.
 *
 * Keyed-merge on `(class, value)`: an equal DataPoint **merges** into the existing entry (keeping
 * `firstRetrieved`, advancing `lastRetrieved`) rather than duplicating. Every batch that changes
 * something advances one monotonic revision, and each entry remembers the revision it was added
 * at and last freshened at; an operator's **watermark** is the revision its last run observed, so
 * the delta since then splits cleanly into added vs updated.
 *
 * @module
 */

import type { AnyDataPoint } from './datapoints.js';
import { DataPointView, identityKey } from './datapoints.js';
import type { CapabilityId, OperatorId } from './ids.js';
import { InvocationDelta } from './operators.js';

/**
 * One held identity: the DataPoint itself and the revisions that touched it.
 *
 * Engine internals are plain objects, not validated models: the loop is their only constructor,
 * so validation would only check the engine against itself — per merge, on the hot path. Mutable
 * on purpose, and private to this module: a re-observation rewrites the entry in place rather
 * than rebuilding the map.
 */
interface Entry {
  dataPoint: AnyDataPoint;
  addedRev: number;
  updatedRev: number;
}

/** What one {@link SessionState.merge} did: the resulting revision, and what was new vs freshened. */
export interface MergeOutcome {
  readonly revision: number;

  /** Identities the batch introduced. */
  readonly added: readonly AnyDataPoint[];

  /** Held identities the batch freshened, as they now stand. */
  readonly updated: readonly AnyDataPoint[];

  /** Whether anything changed at all — precomputed, because every caller asks. */
  readonly changed: boolean;
}

/** Whose caps changed since an operator last ran, as {@link SessionState.deltaFor} takes them. */
export interface DeltaCapabilities {
  /** The capabilities available when this operator last ran. */
  readonly previousCaps: ReadonlySet<CapabilityId>;

  /** The capabilities available now. */
  readonly availableCaps: ReadonlySet<CapabilityId>;
}

/** The session's DataPoints, keyed by identity, plus the revisions the scheduler reads. */
export class SessionState {
  private readonly entries = new Map<string, Entry>();
  private currentRevision = 0;
  private readonly watermarks = new Map<OperatorId, number>();

  /** The current revision; it advances once per batch that changed something. */
  public get revision(): number {
    return this.currentRevision;
  }

  /** Keyed-merge a batch; the revision advances iff something was added or freshened. */
  public merge(dataPoints: Iterable<AnyDataPoint>): MergeOutcome {
    const nextRevision = this.currentRevision + 1;
    const added: AnyDataPoint[] = [];
    const updated: AnyDataPoint[] = [];
    for (const dataPoint of dataPoints) {
      const key = identityKey(dataPoint);
      const existing = this.entries.get(key);
      if (existing === undefined) {
        this.entries.set(key, { dataPoint, addedRev: nextRevision, updatedRev: nextRevision });
        added.push(dataPoint);
      } else if (dataPoint.lastRetrieved.getTime() > existing.dataPoint.lastRetrieved.getTime()) {
        existing.dataPoint = existing.dataPoint.reobserved(dataPoint.lastRetrieved);
        existing.updatedRev = nextRevision;
        updated.push(existing.dataPoint);
      }
    }
    const changed = added.length > 0 || updated.length > 0;
    if (changed) {
      this.currentRevision = nextRevision;
    }
    return Object.freeze({
      revision: this.currentRevision,
      added: Object.freeze(added),
      updated: Object.freeze(updated),
      changed,
    });
  }

  /** A read view over everything held, in the order the identities first arrived. */
  public view(): DataPointView {
    return new DataPointView([...this.entries.values()].map((entry) => entry.dataPoint));
  }

  /** Whether this operator has ever run in this session (it has a watermark). */
  public hasRun(operatorId: OperatorId): boolean {
    return this.watermarks.has(operatorId);
  }

  /** Record the revision an operator's run observed; its next delta starts from there. */
  public advanceWatermark(operatorId: OperatorId, revision: number): void {
    this.watermarks.set(operatorId, revision);
  }

  /** What changed since `operatorId` last ran; a first run presents the whole set as `added`. */
  public deltaFor(operatorId: OperatorId, capabilities: DeltaCapabilities): InvocationDelta {
    const newlyAvailable = [...capabilities.availableCaps].filter((id) => !capabilities.previousCaps.has(id));
    const watermark = this.watermarks.get(operatorId);
    const entries = [...this.entries.values()];
    if (watermark === undefined) {
      return InvocationDelta({
        added: entries.map((entry) => entry.dataPoint),
        updated: [],
        newlyAvailableCaps: newlyAvailable,
        isFirstInvocation: true,
      });
    }
    return InvocationDelta({
      added: entries.filter((entry) => entry.addedRev > watermark).map((entry) => entry.dataPoint),
      // Added at or before the watermark but freshened after it: seen before, seen again since.
      updated: entries
        .filter((entry) => entry.addedRev <= watermark && watermark < entry.updatedRev)
        .map((entry) => entry.dataPoint),
      newlyAvailableCaps: newlyAvailable,
      isFirstInvocation: false,
    });
  }
}
